src/filters/header.rs

```diff
@@ -155,6 +155,78 @@ pub fn exact(
     })
 }
 
+/// Create a `Filter` that requires a header to match the value exactly.
+///
+/// This `Filter` will look for a header with supplied name and the exact
+/// value, otherwise rejects the request.
+///
+/// In comparison to the ([`exact`]) filter, this filter compares the
+/// expected value to the actual value in constant time. Preventing
+/// possible timing attacks.
+///
+/// # Example
+///
+/// ```
+/// // Require `X-API-KEY: abcd` header to be set.
+/// let must_dnt = warp::header::exact_constant_time("X-API-KEY", "abcd");
+/// ```
+pub fn exact_constant_time(
+    name: &'static str,
+    value: &'static str,
+) -> impl Filter<Extract = (), Error = Rejection> + Copy {
+    filter_fn(move |route| {
+        tracing::trace!("exact?({:?}, {:?})", name, value);
+        let route = route
+            .headers()
+            .get(name)
+            .ok_or_else(|| reject::missing_header(name))
+            .and_then(|val| {
+                let incoming_val = match val.to_str() {
+                    Ok(r) => r,
+                    Err(_) => return Err(reject::invalid_header(name)),
+                };
+
+                let len = value.len();
+                let target_len = incoming_val.len();
+
+                // to prevent out of bounds exceptions later.
+                if len == 0 {
+                    return Err(reject::invalid_header(name));
+                };
+
+                // although we can already check if the strings are equal here
+                // by length but due to the timing constraint we still have to check for
+                // equals by byte
+                let mut is_equal = len == target_len;
+
+                let bytes = value.as_bytes();
+                let tgt_bytes = value.as_bytes();
+
+                for n in 0..len {
```

Comment:
Are we sure this will actually compile to constant time? I know there's crates that exist specifically to prevent the compiler from optimizing it...

Reply:
You were absolutely correct, the compiler optimized most stuff out when a hard coded string was provided.

I now rewrote the function to prevent the compiler from doing this.
![2022-10-11_19-21](https://user-images.githubusercontent.com/38077834/195158321-d14e6b4f-3e8b-4e4a-9935-e8f46b10ce7f.png)

Here i build a test program with the --release flag and disassebled it using idafree.
As you can see there is no way past the loop and it has to completely evaluate each byte no matter what.